roles/edpm_nova/scripts/derive_pci_passthrough_whitelist.py

```diff
@@ -0,0 +1,587 @@
+    return pci_dict
+
+
+def get_pci_regex_pattern(config_regex: str, size: int, maxval: int, hex_value: str):
+    if config_regex in [ANY, REGEX_ANY]:
+        config_regex = "[0-9a-fA-F]{%d}" % size
+
+    try:
+        re.compile(config_regex)
+    except re.error:
+        msg = "Invalid regex pattern identified %s" % config_regex
+        raise InvalidConfigException(msg)
+    try:
+        v = int(config_regex, 16)
+    except ValueError:
+        return config_regex
+    if v > maxval:
+        msg = "Invalid pci address"
+        raise InvalidConfigException(msg)
+    return hex_value % v
+
+
+def get_user_regex_from_addrdict(addr_dict):
+    if isinstance(addr_dict, dict):
+        domain_regex = get_pci_regex_pattern(addr_dict['domain'], 4, MAX_DOMAIN, '%04x')
+        bus_regex = get_pci_regex_pattern(addr_dict['bus'], 2, MAX_BUS, '%02x')
+        slot_regex = get_pci_regex_pattern(addr_dict['slot'], 2, MAX_SLOT, '%02x')
+        function_regex = get_pci_regex_pattern(addr_dict['function'], 1, MAX_FUNC, '%1x')
+
+        user_address_regex = '%s:%s:%s.%s' % (domain_regex,
+                                              bus_regex, slot_regex,
+                                              function_regex)
+        return user_address_regex
+    else:
+        return None
+
+
+def get_pci_addr_from_ifname(ifname: str):
+    """Given the device name, returns the PCI address of a device
+    and returns True if the address is in a physical function.
+    """
+    dev_path = os.path.join(_SYS_CLASS_NET_PATH, ifname, "device")
+    if os.path.isdir(dev_path):
+        try:
+            return (os.readlink(dev_path).strip("./"))
+        except OSError:
+            raise
+    return None
+
+
+def get_sriov_configs():
+    configs = []
+    try:
+        with open('/var/lib/os-net-config/sriov_config.yaml') as sriov_config:
+            configs = yaml.safe_load(sriov_config)
+    except IOError:
+        configs = []
+    return configs
+
+
+def get_sriov_nic_partition_pfs(configs):
+    pfs = []
+    for config in configs:
+        device_type = config.get('device_type', None)
+        if device_type and 'vf' in device_type:
+            device = config.get('device', None)
+            if device:
+                dev_name = device.get('name', None)
+                if dev_name and dev_name not in pfs:
+                    pfs.append(dev_name)
+    return pfs
+
+
+def get_sriov_non_nic_partition_pfs(configs):
+    all_pfs = []
+    non_nicp_pfs = []
+    nicp_pfs = get_sriov_nic_partition_pfs(configs)
+
+    for config in configs:
+        device_type = config.get('device_type', None)
+        if device_type and 'pf' in device_type:
+            name = config.get('name', None)
+            if name and name not in all_pfs:
+                all_pfs.append(name)
+    non_nicp_pfs = [x for x in all_pfs if x not in nicp_pfs]
+    return non_nicp_pfs
+
+
+def get_pci_device_info_by_ifname(pci_dir, sub_dir):
```

Comment:
Function can return `None` or a tuple. All invocations of this function must perform a check, or the function must be rewritten so it's not possible.

Reply:
The user input of the derive script is valid pf/vf